src/java/org/apache/cassandra/service/reads/DigestResolver.java

```diff
@@ -44,26 +52,53 @@ public DigestResolver(Keyspace keyspace, ReadCommand command, ConsistencyLevel c
     public void preprocess(MessageIn<ReadResponse> message)
     {
         super.preprocess(message);
-        if (dataResponse == null && !message.payload.isDigestResponse())
-            dataResponse = message.payload;
+        Replica replica = getReplicaFor(message.from);
+        if (dataResponse == null && !message.payload.isDigestResponse() && replica.isFull())
+        {
+            dataResponse = message;
+        }
+        else if (replica.isTransient())
+        {
+            Preconditions.checkArgument(!message.payload.isDigestResponse(), "digest response received from transient replica");
+            hasTransientResponse = true;
+        }
     }
 
     public PartitionIterator getData()
     {
         assert isDataPresent();
-        return UnfilteredPartitionIterators.filter(dataResponse.makeIterator(command), command.nowInSec());
+        int nowInSec = command.nowInSec();
+
+        if (transientResolver == null)
+        {
+            return UnfilteredPartitionIterators.filter(dataResponse.payload.makeIterator(command), nowInSec);
+        }
+        else
+        {
+            return transientResolver.resolve();
+        }
     }
 
     public boolean responsesMatch()
     {
         long start = System.nanoTime();
 
+        TransientResolver resolver = hasTransientResponse
```

Comment:
Maybe a bit nitpicky but maybe name this transientResolver and assign it later with this.transientResolver? I'm not sure if that is more obtuse I guess.